### ISSUE TITLE
LOCATE as an argument to SUBSTRING is silently replaced with 0 when converting DQL to SQL

### ISSUE THREAD

**Issue by datadestroyd** [doctrine/dbal#5974]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 2.11.3

#### Summary

I am trying to select rows that have a column with the same number on both sides of a dash, e.g. "12-12". In SQL I would use the following WHERE clause (Oracle):

```sql
pagination like '%-%' and substr(pagination, 1, instr(pagination, '-') - 1) = substr(pagination, instr(pagination, '-') + 1)
```

I tested it and it works just fine. Now I tried to write a DQL version of this like so:

```dql
a.pagination LIKE '%-%' AND SUBSTRING(a.pagination, 1, LOCATE('-', a.pagination) - 1) = SUBSTRING(a.pagination, LOCATE('-', a.pagination) + 1)
```

However, the generated SQL is this:

```sql
a0_.pagination LIKE '%-%' AND SUBSTR(a0_.pagination, 1, 0) = SUBSTR(a0_.pagination, 0)
```

The LOCATE calls (as well as the +- 1) have been completely removed and replaced with 0. What is going on here?

#### How to reproduce

```php
// Requires an entity with a varchar field called pagination.
var_dump($em->createQueryBuilder()
  ->from(Article::class, "a")
  ->select("a")
  ->where("a.pagination LIKE '%-%'")
  ->andWhere("SUBSTRING(a.pagination, 1, LOCATE('-', a.pagination) - 1) = SUBSTRING(a.pagination, LOCATE('-', a.pagination) + 1)")
  ->getQuery()
  ->getSQL());
```

**Comment by greg0ire** [comment]
When you use `getDQL` instead of `getSQL()` do you see that there is also an issue with the DQL? Trying to understand if this is an issue when parsing, or when transforming the DQL into SQL.

**Comment by datadestroyd** [comment]
`getDQL` returns the correct query.

**Comment by greg0ire** [comment]
Then you might want to debug starting here: https://github.com/doctrine/orm/blob/da9b9de5906802cab1c71f59ffd9fba53ed4011d/lib/Doctrine/ORM/Query/AST/Functions/SubstringFunction.php#L30-L42

**Comment by datadestroyd** [comment]
I've identified the issue. It's actually the fault of the Oracle platform implementation for the SUBSTRING function:

https://github.com/doctrine/dbal/blob/7ef157430c43e41345205ee934f0409854f018bd/src/Platforms/OraclePlatform.php#L60-L67

The position and length arguments are just cast to integers. 🤦‍♂️

**Comment by greg0ire** [comment]
Transferring to the DBAL repository then. A possible next step for you could be to write a failing test, and maybe a fix for that test.

**Comment by MartinMa** [comment]
I added a pull request with a fix and a new test, that fails without the fix.

Please let me know if you need anything else to merge this fix. Thanks.

**Comment by greg0ire** [comment]
Since the fix requires a breaking change, this is fixed on 4.0.x (since https://github.com/doctrine/dbal/pull/3494)
